src/SegmentContentAceessCheck.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+namespace Drupal\openy_gated_content;
+
+use Drupal\Core\Access\AccessResult;
+use Drupal\Core\DependencyInjection\ContainerInjectionInterface;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Messenger\Messenger;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\recurring_events\Entity\EventInstance;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Class SegmentContentAceessCheck.
+ *
+ * @package Drupal\openy_gated_content
+ */
+class SegmentContentAceessCheck implements ContainerInjectionInterface {
+
+  const EDITOR_ROLE = 'virtual_ymca_editor';
+
+  /**
+   * Messenger service.
+   *
+   * @var \Drupal\Core\Messenger\Messenger
+   */
+  protected $messenger;
+
+  /**
+   * Constructor.
+   *
+   * @param Drupal\Core\Messenger\Messenger $messenger
+   *   The entity type manager.
+   */
+  public function __construct(Messenger $messenger) {
+    $this->messenger = $messenger;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('messenger')
+    );
+  }
+
+  /**
+   * Virtual Y check access logic.
+   *
+   * @param \Drupal\Core\Entity\EntityInterface $entity
+   *   Entity to check.
+   * @param string $operation
+   *   Operation type.
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   Current account object.
+   *
+   * @return \Drupal\Core\Access\AccessResultAllowed|\Drupal\Core\Access\AccessResultForbidden|\Drupal\Core\Access\AccessResultNeutral
+   *   Access result object.
+   */
+  public function checkAccess(EntityInterface $entity, $operation, AccountInterface $account) {
+
+    $permissions_config = $this->getEnabledEntities();
+
+    // Our code override's default view permission only.
+    if ($operation !== 'view') {
+      return AccessResult::neutral();
+    }
+
+    $type = $entity->getEntityTypeId();
+
+    // Skip other entities.
+    if (!in_array($type, array_keys($permissions_config))) {
+      return AccessResult::neutral();
+    }
+
+    if (array_key_exists($type, $permissions_config)) {
+      $bundle = $entity->bundle();
+
+      if (in_array($bundle, $permissions_config[$type])) {
+
+        $account_roles = $account->getRoles();
+
+        // Use Drupal permissions for administrators and editors.
+        if (
+          in_array(self::EDITOR_ROLE, $account_roles)
+          || in_array('administrator', $account_roles)
+        ) {
+          return AccessResult::neutral();
+        }
+
+        $content_access_mask = $entity->get('field_vy_permission')->getValue();
+
+        // For Eventinstance we have to check parent as well.
+        if (($entity instanceof EventInstance) && empty($content_access_mask)) {
+          $eventSeriesEntity = $entity->getEventSeries();
+          $content_access_mask = $eventSeriesEntity->get('field_vy_permission')
```

Comment:
I think we need to check `field_vy_permission` from `EventInstance` too. For example, if `field_vy_permission` not empty in `EventInstance` - use this value, otherwise - use value from parent.

Reply:
at my code it does exactly the same.
1. If event instance has its own field value - we use it for check.
2. If eventinstance permissions value is empty - we are checking it at event series.

Reply:
Oh. sorry, missed this, ok